simulator/network.py

```diff
@@ -47,21 +64,60 @@ def send_unicast(self, from_addr, to_addr, msg):
                         to_addr, from_addr))
 
     def send_broadcast(self, from_addr, msg):
-        print('Message Broadcast from {} - {}'.format(from_addr,msg))
-        for to_addr in self.addr_list:
-            if(to_addr <= 0):
-                print('{} address not found (msg from {})'.format(
-                    to_addr, from_addr))
-                yield self.env.timeout(0)
-            else: 
+        print(str(self.env.now) + ' :: ' + 'Message Broadcast from {} - {}'.format(from_addr,msg))
+        with self.node_list_access.request(priority=0) as nl_access:
+            yield nl_access
+            for addr in range(len(self.node_list)):
+                to_addr = self.node_list[addr]['address']
                 msg_envelope2 = [from_addr, to_addr, msg]
                 with self.channel.request() as rec:
                     yield rec
-                    node = self.node_map[to_addr]
-                    if node:
+                    node = self.node_list[addr]['node']
+                    if node:                # Checando se o nodo ainda faz parte da rede
                         node.recieve(msg_envelope2)
                         yield self.env.timeout(self.latency)
-                        print('Broadcast:'+ str(msg_envelope2))
+                        print(str(self.env.now) + ' :: ' + 'Broadcast:'+ str(msg_envelope2))
                     else:
                         print('{} address not found (msg from {})'.format(
-                            to_addr, from_addr))
\ No newline at end of file
+                            to_addr, from_addr))
+
+    def find_next_available(self):
+        addr = (self.next_available_address + 1) % self.max_hosts
+        for i in range(self.max_hosts - 1):
+            if self.addr_list[addr]['node'] == None:
+                self.next_available_address = addr
+                self.full_capacity = False
+                return
+            addr = (addr + 1) % self.max_hosts
+        self.next_available_address = 0            # Se chegarmos até aqui, então não há endereços disponíveis
+        self.full_capacity = True
+
+    def renew(self, address):
+        if self.addr_list[address]['node']:
+            self.addr_list[address]['lease'] = self.env.now + self.default_lease_time
+            print(str(self.env.now) + ' :: ' + 'Lease renewed for address {}'.format(address))
+
+    def check_lease(self):
+        with self.node_list_access.request(priority=1) as nl_access:
+            yield nl_access
+            nodes_to_delete = []
+            for i in range(len(self.node_list)):
+                if self.node_list[i]['time'] <= self.env.now:
+                    nodes_to_delete.append(i)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider using enumerate instead of iterating with range and len](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5428630)